File: occupancy-manager-java/src/test/java/com/occupancy/manager/controller/OccupancyManagerControllerTest.java
package com.occupancy.manager.controller;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.when;

import com.occupancy.manager.factory.OccupancyUsageRequestFactory;
import com.occupancy.manager.factory.OccupancyUsageResponseFactory;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import com.occupancy.manager.dto.OccupancyUsageRequest;
import com.occupancy.manager.service.OccupancyManagerService;

/**
 * 
 * @author Ramachandra
 *
 */
class OccupancyManagerControllerTest {

	@InjectMocks
	OccupancyManagerController occupancyManagerController;

	@Mock
	private OccupancyManagerService occupancyManagerService;

	@BeforeEach
	public void init() {
		MockitoAnnotations.openMocks(this);
	}

	@Test
	void testcalculateOccupancyFor3Premium3EconomyRooms() {
		when(occupancyManagerService.calculateOccupancy(any(OccupancyUsageRequest.class)))
				.thenReturn(OccupancyUsageResponseFactory.createOccupancyUsageResponseFor3Premium3EconomyRooms());
		final var occupancyUsageResponse = occupancyManagerController
				.calculateOccupancy(OccupancyUsageRequestFactory.createThreePremiumAndThreeEconomyRooms());
		assertEquals("3 (EUR 738)", occupancyUsageResponse.getBody().getPremiumUsage(),
				"Expected the premium usage values to match");
		assertEquals("3 (EUR 167)", occupancyUsageResponse.getBody().getEconomyUsage(),
				"Expected the economy usage values to match");
	}
	
	@Test
	void testcalculateOccupancyFor7Premium1EconomyRoom() {
		when(occupancyManagerService.calculateOccupancy(any(OccupancyUsageRequest.class)))
				.thenReturn(OccupancyUsageResponseFactory.createOccupancyUsageResponseFor7Premium1EconomyRoom());
		final var occupancyUsageResponse = occupancyManagerController
				.calculateOccupancy(OccupancyUsageRequestFactory.createSevenPremiumAndOneEconomyRoom());
		assertEquals("7 (EUR 1153)", occupancyUsageResponse.getBody().getPremiumUsage(),
				"Expected the premium usage values to match");
		assertEquals("1 (EUR 45)", occupancyUsageResponse.getBody().getEconomyUsage(),
				"Expected the economy usage values to match");
	}
	
	@Test
	void testcalculateOccupancyFor2Premium7EconomyRooms() {
		when(occupancyManagerService.calculateOccupancy(any(OccupancyUsageRequest.class)))
				.thenReturn(OccupancyUsageResponseFactory.createOccupancyUsageResponseFor2Premium7EconomyRooms());
		final var occupancyUsageResponse = occupancyManagerController
				.calculateOccupancy(OccupancyUsageRequestFactory.createTwoPremiumAndSevenEconomyRooms());
		assertEquals("2 (EUR 583)", occupancyUsageResponse.getBody().getPremiumUsage(),
				"Expected the premium usage values to match");
		assertEquals("4 (EUR 189)", occupancyUsageResponse.getBody().getEconomyUsage(),
				"Expected the economy usage values to match");
	}
	
	@Test
	void testcalculateOccupancyFor7Premium5EconomyRooms() {
		when(occupancyManagerService.calculateOccupancy(any(OccupancyUsageRequest.class)))
				.thenReturn(OccupancyUsageResponseFactory.createOccupancyUsageResponseFor7Premium5EconomyRooms());
		final var occupancyUsageResponse = occupancyManagerController
				.calculateOccupancy(OccupancyUsageRequestFactory.createSevenPremiumAndFiveEconomyRooms());
		assertEquals("6 (EUR 1054)", occupancyUsageResponse.getBody().getPremiumUsage(),
				"Expected the premium usage values to match");
		assertEquals("4 (EUR 189)", occupancyUsageResponse.getBody().getEconomyUsage(),
				"Expected the economy usage values to match");
	}
}

File: occupancy-manager-kotlin/src/test/kotlin/com/occupancy/manager/controller/OccupancyManagerControllerTest.kt
package com.occupancy.manager.controller

import com.occupancy.manager.dto.OccupancyUsageResponse
import com.occupancy.manager.factory.OccupancyUsageRequestFactory
import com.occupancy.manager.factory.OccupancyUsageResponseFactory
import com.occupancy.manager.service.OccupancyManagerService
import io.mockk.every
import io.mockk.mockk
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.springframework.http.ResponseEntity

/**
 *
 * @author Ramachandra
 */
internal class OccupancyManagerControllerTest() {

    private val occupancyManagerService = mockk<OccupancyManagerService>()

    private var occupancyManagerController = OccupancyManagerController(occupancyManagerService)

    @Test
    fun testCalculateOccupancyFor3Premium3EconomyRooms() {
        val request = OccupancyUsageRequestFactory.createThreePremiumAndThreeEconomyRooms()
        every { occupancyManagerService.calculateOccupancy(request) } returns
                OccupancyUsageResponseFactory.createOccupancyUsageResponseFor3Premium3EconomyRooms()
        val occupancyUsageResponse: ResponseEntity<OccupancyUsageResponse> = occupancyManagerController
            .calculateOccupancy(request)
        assertEquals(
            "3 (EUR 738)", occupancyUsageResponse.body!!.premiumUsage,
            "Expected the premium usage values to match"
        )
        assertEquals(
            "3 (EUR 167)", occupancyUsageResponse.body!!.economyUsage,
            "Expected the economy usage values to match"
        )
    }

    @Test
    fun testCalculateOccupancyFor7Premium1EconomyRoom() {
        val request = OccupancyUsageRequestFactory.createSevenPremiumAndOneEconomyRoom()
        every { occupancyManagerService.calculateOccupancy(request) } returns
                OccupancyUsageResponseFactory.createOccupancyUsageResponseFor7Premium1EconomyRoom()
        val occupancyUsageResponse: ResponseEntity<OccupancyUsageResponse> = occupancyManagerController
            .calculateOccupancy(request)
        assertEquals(
            "7 (EUR 1153)", occupancyUsageResponse.body!!.premiumUsage,
            "Expected the premium usage values to match"
        )
        assertEquals(
            "1 (EUR 45)", occupancyUsageResponse.body!!.economyUsage,
            "Expected the economy usage values to match"
        )
    }

    @Test
    fun testCalculateOccupancyFor2Premium7EconomyRooms() {
        val request = OccupancyUsageRequestFactory.createTwoPremiumAndSevenEconomyRooms()
        every { occupancyManagerService.calculateOccupancy(request) } returns
                OccupancyUsageResponseFactory.createOccupancyUsageResponseFor2Premium7EconomyRooms()
        val occupancyUsageResponse: ResponseEntity<OccupancyUsageResponse> = occupancyManagerController
            .calculateOccupancy(request)
        assertEquals(
            "2 (EUR 583)", occupancyUsageResponse.body!!.premiumUsage,
            "Expected the premium usage values to match"
        )
        assertEquals(
            "4 (EUR 189)", occupancyUsageResponse.body!!.economyUsage,
            "Expected the economy usage values to match"
        )
    }

    @Test
    fun testCalculateOccupancyFor7Premium5EconomyRooms() {
        val request = OccupancyUsageRequestFactory.createSevenPremiumAndFiveEconomyRooms()
        every { occupancyManagerService.calculateOccupancy(request) } returns
                OccupancyUsageResponseFactory.createOccupancyUsageResponseFor7Premium5EconomyRooms()
        val occupancyUsageResponse: ResponseEntity<OccupancyUsageResponse> = occupancyManagerController
            .calculateOccupancy(request)
        assertEquals(
            "6 (EUR 1054)", occupancyUsageResponse.body!!.premiumUsage,
            "Expected the premium usage values to match"
        )
        assertEquals(
            "4 (EUR 189)", occupancyUsageResponse.body!!.economyUsage,
            "Expected the economy usage values to match"
        )
    }
}